File: PRACTICA_JAVA/src/irrgarten/Shield.java
package irrgarten;

/**
 * Represents a shield that provides protection in the Irrgarten game.
 * Shields have a protection value and a number of uses before they are discarded.
 * 
 * @author marcosbslinux
 * @version 1.0
 */
public class Shield {
    /** Protection value of the shield */
    private float protection;
    
    /** Number of remaining uses */
    private int uses;
    
    /**
     * Constructs a new Shield with specified protection and uses
     * @param p Protection value
     * @param u Number of uses
     */
    public Shield(float p, int u){
        protection=p;
        uses=u;
    }
    
    /**
     * Uses the shield to provide protection
     * @return Protection value if shield has uses remaining, 0 otherwise
     */
    public float protect(){
        float defense=0;
        
        if(uses>0){
            --uses;
            defense=protection;
        }
        return defense;
    }
    /**
     * Generates string representation of shield's state
     * @return String showing protection and remaining uses
     */
    @Override
    public String toString(){
        return "S[" + Float.toString(protection) + ", " + Integer.toString(uses) + "]";
    }
    /**
     * Determines if shield should be discarded based on remaining uses
     * @return true if shield should be discarded, false otherwise
     */
    public boolean discard(){
        return Dice.discardElement(uses);
    }
}

File: PRACTICA_RUBY/shield.rb
#encoding: utf-8
module Irrgarten
  # Represents a shield in the Irrgarten game
  # Manages shield protection and remaining uses
  class Shield
    # Initializes a new shield with specified protection and uses
    # @param p [Integer] protection value of the shield
    # @param u [Integer] number of remaining uses
    def initialize(p, u)
      @protection=p
      @uses=u
    end

    # Calculates defense value from the shield
    # Decreases uses by 1 if shield has remaining uses
    # @return [Integer] protection value if shield has uses, 0 otherwise
    def protect
      defense=0
      if @uses>0
        @uses -= 1
        defense=@protection
      end
      defense
    end

    # Returns string representation of shield state
    # @return [String] formatted shield description showing protection and uses
    def to_s
      "S[#{@protection}, #{@uses}]"
    end

    # Determines if shield should be discarded
    # Uses dice roll to decide based on remaining uses
    # @return [Boolean] true if shield should be discarded
    def discard
      Dice.discard_element(@uses)
    end
  end
end
